File: examples/60-factorial.pas
Program Factorial;

Const
  Trace = False;

Function RecursiveFactorial(N: Unsigned): Unsigned;
Begin
  If N <= 1 Then
    Begin
      Result := 1;
      If Trace Then
        WriteLn('RecursiveFactorial(', N, ') = ', Result);
    End
  Else
    Begin
      Result := N * RecursiveFactorial(N -1);
      If Trace Then
        WriteLn('RecursiveFactorial(', N, ') = ', Result);
    End;
End;

Function IterativeFactorial(N: Unsigned): Unsigned;
Var
  I, F: Unsigned;
Begin
  If N <= 1 Then
    Begin
      If Trace Then
        WriteLn('IterativeFactorial: I=', I, ', F=', F);
      F := 1;
    End
  Else
    Begin
      F := 1;
      For I := 2 To N Do
      Begin
        If Trace Then
          WriteLn('IterativeFactorial: I=', I, ', F=', F);
        F := F * I;
      End;
    End;
  // IterativeFactorial := F;
  Result := F;
End;

Var
  N, Recursive, iterative: Unsigned;
Begin
  // Repeat
  //   Write('N=');
  //   ReadLn(N);
  // Until N > 0;
  // N := 31;
  WriteLn('Factorial:');
  For N := 0 To 15 Do
  Begin
    Recursive := RecursiveFactorial(N);
    // WriteLn(' - Recursive: ', N, '! = ', Recursive);
    iterative := IterativeFactorial(N);
    // WriteLn(' - Iterative: ', N, '! = ', iterative);
    WriteLn(' - ', N, '! = ', Recursive, ' = ', iterative);
  End;
End.

File: examples/60-factorial.c
// Converted with https://www.codeconvert.ai/pascal-to-c-converter

#include <stdio.h>

int recursive_factorial(int n) {
    if (n <= 1) {
        return 1;
    } else {
        return n * recursive_factorial(n - 1);
    }
}

int iterative_factorial(int n) {
    int i, f;
    if (n <= 1) {
        f = 1;
    } else {
        f = 1;
        for (i = 2; i <= n; i++) {
            f = f * i;
        }
    }
    return f;
}

int main() {
    int n;
    do {
        printf("N=");
        scanf("%d", &n);
    } while (n <= 0);
    
    printf("Recursive: %d! = %d\n", n, recursive_factorial(n));
    printf("Iterative: %d! = %d\n", n, iterative_factorial(n));
    
    return 0;
}
